include/oneapi/tbb/collaborative_once_flag.h

```diff
@@ -0,0 +1,126 @@
+        return cancellation_status ? canceled : complete;
+    }
+
+    /*! A call to wait() potentially participates in the work of run_and_wait().
+     * If run_and_wait() throws or is canceled, this returns canceled.
+     */
+    task_group_status wait() {
+        bool cancellation_status = false;
+        auto f = [&]() {
+            detail::try_call([&] {
+                detail::d1::wait(my_wait_context, my_context);
+            }).on_completion([&] {
+                cancellation_status = my_context.is_group_execution_cancelled();
+            });
+        };
+        my_arena.execute(f);
+        return cancellation_status ? canceled : complete;
+    }
+
+};
+
+} // namespace detail
+
+class canceled_exception : public std::exception {};
+
+// The concurrent_once_flag idea may be added to tbb at some point: oneapi-src/oneTBB#267
+class collaborative_once_flag : detail::d0::no_copy {
+    enum class state { uninitialized, in_progress, done };
+    std::atomic<state> my_state = state::uninitialized;
+    std::shared_ptr<detail::run_once> my_runner = nullptr;
+
+    template <typename Fn>
+    friend void collaborative_call_once(collaborative_once_flag& flag, Fn&& f);
+
+    template <typename Fn>
+    void do_collaborative_call_once(Fn&& f) {
+        while (my_state.load(std::memory_order_acquire) !=
+               state::done) { // Loop until someone else finishes or it's Done.
+            state expected = state::uninitialized;
+            if (my_state.compare_exchange_weak(expected, state::in_progress)) {
+                // We won the race to be the primary thread.
+                auto runner = std::make_shared<detail::run_once>();
+                std::atomic_store_explicit(&my_runner, runner, std::memory_order_release);
+                try {
+                    tbb::task_group_status wait_result = runner->run_and_wait(f);
+                    if (wait_result == tbb::task_group_status::canceled) {
+                        throw canceled_exception(); // Translate cancelation to an exception.
+                    }
+                } catch (...) {
+                    // Re-set and re-throw:
+                    std::atomic_store_explicit(&my_runner, std::shared_ptr<detail::run_once>(nullptr),
+                                               std::memory_order_release);
+                    my_state.store(state::uninitialized, std::memory_order_release);
+                    throw;
+                }
+                // Let go, mark done, and return:
+                std::atomic_store_explicit(&my_runner, std::shared_ptr<detail::run_once>(nullptr),
+                                           std::memory_order_release);
+                my_state.store(state::done, std::memory_order_release);
+                return;
+            } else {
+                // Get our own pointer so it can't go away while we aren't looking.
+                if (auto runner = std::atomic_load_explicit(&my_runner, std::memory_order_acquire)) {
```

Comment:
it seems that we are busy waiting for runner publication or for `my_state` completion that requires some sort of backoff.

Reply:
`tbb::internal::atomic_backoff`? That seems too useful to hide in `internal`.